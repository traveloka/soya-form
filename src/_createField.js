import React from 'react';
import PropTypes from 'prop-types';
import hoistStatics from 'hoist-non-react-statics';
import { formShape } from './_propTypes';
import {
  createValidate,
  createValidateAsync,
  getDisplayName,
  throwAndLogError,
} from './_utils';

export default Component => hoistStatics(class extends React.Component {
  static displayName = getDisplayName('CreateField', Component);

  static propTypes = {
    form: formShape.isRequired,
    name: PropTypes.arrayOf(PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.string,
    ])).isRequired,
    isFormEnabled: PropTypes.bool.isRequired,
    isFormSubmitting: PropTypes.bool.isRequired,
    isDisabled: PropTypes.bool.isRequired,
    errorMessages: PropTypes.array.isRequired,
    isFieldEnabled: PropTypes.bool.isRequired,
    isValidating: PropTypes.bool.isRequired,
    touched: PropTypes.bool.isRequired,
    value: PropTypes.any, // eslint-disable-line react/forbid-prop-types
    setDefaultValue: PropTypes.func.isRequired,
    changeHandlers: PropTypes.arrayOf(PropTypes.func),
    changeValidators: PropTypes.arrayOf(PropTypes.func),
    createHandleChange: PropTypes.func.isRequired,
    asyncValidators: PropTypes.arrayOf(PropTypes.func),
    submitValidators: PropTypes.arrayOf(PropTypes.func),
    addErrorMessages: PropTypes.func.isRequired,
    mergeFields: PropTypes.func.isRequired,
    toggleLock: PropTypes.func.isRequired,
    preventRemoveField: PropTypes.bool
  };

  static defaultProps = {
    changeHandlers: [],
    changeValidators: [],
    asyncValidators: [],
    submitValidators: [],
    parseValue: value => value,
  };

  constructor(props, context) {
    super(props, context);
    this.__form = props.form;
    this.__asyncValidators = [];
    this.__changeValidators = [];
    this.__submitValidators = [];
  }

  componentWillMount() {
    const handleValidateAll = this.createHandleValidateAll();
    this.__form.regField(this.props.name, handleValidateAll);
  }

  componentWillUnmount() {
    if (!this.props.preventRemoveField) {
      this.__form.unregField(this.props.name);
    }
  }

  addErrorMessages = errorMessages => this.props.addErrorMessages(this.props.name)(errorMessages);

  mergeFields = object => this.props.mergeFields(this.props.name)(object);

  setDefaultValue = value => this.props.setDefaultValue(this.props.name)(value);

  toggleLock = shouldLock => this.props.toggleLock(this.props.name)(shouldLock);

  lock = () => this.toggleLock(true);

  unlock = () => this.toggleLock(false);

  resolveAsyncValidation = errorMessages => {
    this.addErrorMessages(errorMessages);
    this.unlock();
    return errorMessages;
  };

  rejectAsyncValidation = error => {
    this.unlock();
    throwAndLogError(error);
  };

  createHandleChange = validators => this.props.createHandleChange(this.props.name)(validators);

  createHandleAsyncValidation = (validators, asyncValidators) => () => {
    const errorMessages = createValidate(validators)(this.props.value);
    if (errorMessages.length > 0) {
      return Promise.resolve(false);
    }

    if (asyncValidators.length === 0) {
      return Promise.resolve(true);
    }

    this.lock();
    return createValidateAsync(asyncValidators)(this.props.value)
      .then(this.resolveAsyncValidation)
      .catch(this.rejectAsyncValidation);
  };

  createHandleValidateAll = () => () => {
    const validators = [
      ...this.props.changeValidators,
      ...this.__changeValidators,
    ];
    const asyncValidators = [
      ...this.props.asyncValidators,
      ...this.__asyncValidators,
      ...this.props.submitValidators,
      ...this.__submitValidators,
    ];
    const errorMessages = createValidate(validators)(this.props.value);
    const createResult = isValid => {
      const { parseValue } = this.props;

      if (typeof parseValue !== 'function') {
        throw Error('ParseValue should be function');
      }

      return {
        isValid,
        value: parseValue(this.props.value),
        name: this.props.name,
      };
    };
    if (errorMessages.length > 0) {
      this.mergeFields({ errorMessages });
      return Promise.resolve(createResult(false));
    }

    if (asyncValidators.length === 0) {
      return Promise.resolve(createResult(true));
    }

    this.lock();
    return createValidateAsync(asyncValidators)(this.props.value)
      .then(this.resolveAsyncValidation)
      .then(errorMessages => createResult(errorMessages.length === 0))
      .catch(this.rejectAsyncValidation);
  };

  registerAsyncValidators(asyncValidators) {
    this.__asyncValidators = this.__asyncValidators.concat(asyncValidators);
  }

  registerChangeValidators(changeValidators) {
    this.__changeValidators = this.__changeValidators.concat(changeValidators);
  }

  registerSubmitValidators(submitValidators) {
    this.__submitValidators = this.__submitValidators.concat(submitValidators);
  }

  render() {
    const {
      changeValidators,
      asyncValidators,
      ...props
    } = this.props;
    delete props.createHandleChange;
    delete props.setDefaultValue;
    delete props.changeHandlers;
    delete props.submitValidators;
    delete props.addErrorMessages;
    delete props.mergeFields;
    delete props.toggleLock;

    const validators = [
      ...changeValidators,
      ...this.__changeValidators,
    ];
    props.setDefaultValue = this.setDefaultValue;
    props.handleChange = this.createHandleChange(validators);
    props.handleAsyncValidation = this.createHandleAsyncValidation(validators, [
      ...asyncValidators,
      ...this.__asyncValidators,
    ]);
    props.registerChangeValidators = this.registerChangeValidators;
    props.registerAsyncValidators = this.registerAsyncValidators;
    props.registerSubmitValidators = this.registerSubmitValidators;

    return <Component {...props} />;
  }
}, Component);
