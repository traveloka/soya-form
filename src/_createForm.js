import React from 'react';
import PropTypes from 'prop-types';
import hoistStatics from 'hoist-non-react-statics';
import {
  addErrorMessages,
  clear,
  clearErrorMessages,
  mergeFields,
  setDefaultValue,
  setDefaultValues,
  setErrorMessages,
  setFormEnabledState,
  setFormIsSubmittingState,
  setValue,
  setValues,
  initField,
} from './_actions';
import { CONTEXT_NAME } from './_constants';
import { formShape } from './_propTypes';
import {
  getDisplayName,
  throwAndLogError,
} from './_utils';

export const __createForm = (fields, fieldNames) => (formId, dispatch, preventRemoveField) => {
  const __fields = fields || {};
  const __fieldNames = fieldNames || [];
  const actionCreators = {
    lockSubmission: () => setFormIsSubmittingState(formId, true),
    unlockSubmission: () => setFormIsSubmittingState(formId, false),
    disable: () => setFormEnabledState(formId, false),
    enable: () => setFormEnabledState(formId, true),
    disableField(fieldName) {
      return this.disableFields([fieldName]);
    },
    enableField(fieldName) {
      return this.enableFields([fieldName]);
    },
    disableFields: fieldNames => mergeFields(formId, fieldNames.map(fieldName => ({
      fieldName,
      object: { isEnabled: false },
    }))),
    enableFields: fieldNames => mergeFields(formId, fieldNames.map(fieldName => ({
      fieldName,
      object: { isEnabled: true },
    }))),
    setValue: (fieldName, value) => setValue(formId, fieldName, value),
    setValues: values => setValues(formId, values),
    setDefaultValue: (fieldName, value) => setDefaultValue(formId, fieldName, value),
    setDefaultValues: values => setDefaultValues(formId, values),
    setErrors: (fieldName, errorMessages) => setErrorMessages(formId, fieldName, errorMessages),
    addErrors: errorMessages => addErrorMessages(formId, errorMessages),
    clearForm: () => clear(formId),
    clearErrors: fieldNames => clearErrorMessages(formId, fieldNames),
  };

  return {
    ...Object.keys(actionCreators).reduce((methods, name) => ({
      ...methods,
      [name]: (...args) => dispatch(actionCreators[name](...args)),
    }), {}),
    getFormId: () => formId,
    regField: (fieldNames, validateAll) => {
      dispatch(initField(formId, fieldNames));
      __fields[fieldNames] = { validateAll };
      __fieldNames.push(fieldNames);
    },
    unregField: fieldNames => {
      if (!preventRemoveField) {
        delete __fields[fieldNames];
        const index = __fieldNames.findIndex(
          __fieldNames => (
            __fieldNames.length === fieldNames.length &&
            __fieldNames.every((name, i) => name === fieldNames[i])
          ));

        if (index !== -1) {
          __fieldNames.splice(index, 1);
        }
      }
    },
    validateAll: () => {
      const promises = Object.keys(__fields).map(fieldName => __fields[fieldName].validateAll());
      return Promise.all(promises)
        .then(results => {
          let values = {};
          if (
            results.length > 0 &&
            results[0].name.length > 0 &&
            typeof results[0].name[0] === 'number'
          ) {
            values = [];
          }
          return results.reduce((newResult, result) => ({
            isValid: newResult.isValid && result.isValid,
            values: result.name.reduce((value, name, index) => {
              if (index === result.name.length - 1) {
                value[name] = result.value;
                return newResult.values;
              }
              if (!value.hasOwnProperty(name)) {
                value[name] = typeof result.name[index + 1] === 'number' ? [] : {};
              }
              return value[name];
            }, newResult.values),
          }), {
            isValid: true,
            values,
          });
        })
        .catch(throwAndLogError);
    },
    submit(submitFunc, validationFunc) {
      dispatch(clearErrorMessages(formId, __fieldNames));
      return this.validateAll()
        .then(result => {
          if (!result.isValid || !validationFunc) {
            submitFunc(result);
            return;
          }

          return Promise.resolve(validationFunc(result.values))
            .then(validationResult => {
              if (typeof validationResult === 'object' && !validationResult.isValid) {
                dispatch(addErrorMessages(formId, validationResult.errorMessages));
                result.isValid = false;
              }
              submitFunc(result);
            })
            .catch(error => {
              result.isValid = false;
              submitFunc(result);
              throwAndLogError(error);
            });
        })
        .catch(error => {
          throwAndLogError(error);
        });
    },
  };
};

export const _createForm = __createForm();

export default (formId, preventRemoveField) => Component => hoistStatics(class extends React.Component {
  static displayName = getDisplayName('CreateForm', Component);

  static childContextTypes = {
    [CONTEXT_NAME]: formShape.isRequired,
  };

  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    init: PropTypes.func.isRequired,
    formId: PropTypes.string,
  };

  getChildContext = () => ({
    [CONTEXT_NAME]: this.__form,
  });

  constructor(props, context) {
    super(props, context);
    let newFormId = formId || props.formId;
    if (typeof newFormId === 'function') {
      newFormId = newFormId(props);
    }
    this.__form = _createForm(newFormId, props.dispatch, preventRemoveField);
  }

  componentWillMount() {
    this.props.init();
  }

  render() {
    const props = { ...this.props };
    delete props.dispatch;
    delete props.formId;
    props.form = this.__form;
    return <Component {...props} />;
  }
}, Component);
