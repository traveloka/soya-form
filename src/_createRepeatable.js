import React from 'react';
import PropTypes from 'prop-types';
import hoistStatics from 'hoist-non-react-statics';
import { formShape } from './_propTypes';
import {
  createGetName,
  getDisplayName,
} from './_utils';

export default Component => hoistStatics(class extends React.Component {
  static displayName = getDisplayName('CreateRepeatable', Component);

  static propTypes = {
    form: formShape.isRequired, // eslint-disable-line react/no-unused-prop-types
    length: PropTypes.number.isRequired,
    maxLength: PropTypes.number, // eslint-disable-line react/no-unused-prop-types
    minLength: PropTypes.number, // eslint-disable-line react/no-unused-prop-types
    name: PropTypes.arrayOf(PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.string,
    ])).isRequired,
    addListItem: PropTypes.func.isRequired,
    removeListItem: PropTypes.func.isRequired,
    reorderListItem: PropTypes.func.isRequired,
    reorderListItemDec: PropTypes.func.isRequired,
    reorderListItemInc: PropTypes.func.isRequired,
    tag: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  };

  static defaultProps = {
    minLength: 0,
    tag: 'div'
  };

  addListItem = () => {
    const names = this.props.name.slice();
    if (typeof names[names.length - 1] === 'string') {
      names.push(names.length);
    }
    this.props.addListItem(names)();
  };

  removeListItem = index => this.props.removeListItem(this.props.name)(index);

  reorderListItem = (index, targetIndex) => this.props.reorderListItem(this.props.name)(index, targetIndex);

  reorderListItemDec = (index, amount) => this.props.reorderListItemDec(this.props.name)(index, amount);

  reorderListItemInc = (index, amount) => this.props.reorderListItemInc(this.props.name)(index, amount);

  render() {
    const {
      tag: Tag,
      ...props
    } = this.props;

    return (
      <Tag>
        {new Array(this.props.length).fill(null).map((_, index) => (
          <Component
            {...props}
            index={index}
            key={index}
            name={createGetName(this.props.name, index)}
            totalItem={this.props.length}
            addListItem={this.addListItem}
            removeListItem={this.removeListItem}
            reorderListItem={this.reorderListItem}
            reorderListItemDec={this.reorderListItemDec}
            reorderListItemInc={this.reorderListItemInc}
          />
        ))}
      </Tag>
    );
  }
}, Component);
