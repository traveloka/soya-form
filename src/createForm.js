import { compose } from 'redux';
import { connect } from 'react-redux';
import { applyReducers } from 'soya-next/redux';
import { initForm } from './_actions';
import reducers from './_reducers';
import createForm from './_createForm';

const createMapDispatchToProps = formId => (dispatch, props) => ({
  dispatch,
  init() {
    let newFormId = formId;
    if (typeof formId === 'function') {
      newFormId = formId(props);
    }
    dispatch(initForm(newFormId));
  },
});

export default (formId, preventRemoveField) => compose(
  applyReducers(reducers),
  connect(null, createMapDispatchToProps(formId)),
  createForm(formId, preventRemoveField)
);
