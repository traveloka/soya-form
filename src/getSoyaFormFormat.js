import isArray from "lodash/isArray";
import isObject from "lodash/isObject";
import isPlainObject from "lodash/isPlainObject";

export const transformObject = (operand, parentKeys) => {
  return Object.keys(operand).reduce((carry, key) => {
    // primitive value see: https://github.com/lodash/lodash/issues/1406
    if (!isObject(operand[key])) {
      return [
        ...carry,
        { fieldName: [...parentKeys, key], value: operand[key] }
      ];
    }

    // Array value
    if (isArray(operand[key])) {
      return [...carry, ...transformArray(operand[key], [...parentKeys, key])];
    }

    // Object value
    if (isPlainObject(operand[key])) {
      return [...carry, ...transformObject(operand[key], [...parentKeys, key])];
    }

    return carry;
  }, []);
};

export const transformArray = (operand, parentKeys) => {
  return operand.reduce((carry, arrayItem, currIdx) => {
    // primitive value see: https://github.com/lodash/lodash/issues/1406
    if (!isObject(arrayItem)) {
      return [
        ...carry,
        { fieldName: [...parentKeys, currIdx], value: arrayItem }
      ];
    }

    // Object value
    if (isPlainObject(arrayItem)) {
      return [
        ...carry,
        ...transformObject(arrayItem, [...parentKeys, currIdx])
      ];
    }

    return carry;
  }, []);
};

export const transform = serverResponse => {
  return Object.keys(serverResponse).reduce((carry, key) => {
    // primitive value see: https://github.com/lodash/lodash/issues/1406
    if (!isObject(serverResponse[key])) {
      return [...carry, { fieldName: key, value: serverResponse[key] }];
    }

    // Array value
    if (isArray(serverResponse[key])) {
      return [...carry, ...transformArray(serverResponse[key], [key])];
    }

    // Object value
    if (isPlainObject(serverResponse[key])) {
      return [...carry, ...transformObject(serverResponse[key], [key])];
    }

    return carry;
  }, []);
};
