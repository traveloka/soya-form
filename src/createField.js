import { compose } from 'redux';
import { connect } from 'react-redux';
import { applyReducers } from 'soya-next/redux';
import reducers, { initialState } from './_reducers';
import {
  addErrorMessages,
  mergeFields,
  setDefaultValue,
  setIsValidating,
  setValue,
} from './_actions';
import createSelector from './_selectors';
import {
  createValidate,
  getFieldNames,
} from './_utils';
import withForm from './withForm';
import _createField from './_createField';

/* eslint-disable complexity */
const mapStateToProps = (state, props) => {
  const formId = props.form.getFormId();
  const fieldNames = getFieldNames(props.name);
  const selector = createSelector(state, formId);
  const form = selector.getForm() || initialState.form;
  const field = selector.getField(fieldNames) || initialState.field;

  return {
    name: fieldNames,
    isFormEnabled: form.isEnabled,
    isFormSubmitting: form.isSubmitting,
    isDisabled: field.isValidating || !form.isEnabled || !field.isEnabled || form.isSubmitting,
    errorMessages: field.errorMessages,
    isFieldEnabled: field.isEnabled,
    isValidating: field.isValidating,
    touched: field.touched,
    value: typeof (field.value) === 'undefined' || field.value === null ? '' : field.value
  };
};
/* eslint-enable complexity */

const mapDispatchToProps = (dispatch, props) => ({
  mergeFields: fieldNames => object => {
    dispatch(mergeFields(props.form.getFormId(), [{
      fieldName: fieldNames,
      object,
    }]));
  },
  toggleLock: fieldNames => shouldLock => {
    dispatch(setIsValidating(props.form.getFormId(), { [fieldNames]: shouldLock }));
  },
  addErrorMessages: fieldNames => errorMessages => {
    dispatch(addErrorMessages(props.form.getFormId(), [{
      fieldName: fieldNames,
      errorMessages,
    }]));
  },
  setDefaultValue: fieldNames => value => {
    dispatch(setDefaultValue(props.form.getFormId(), fieldNames, value));
  },
  createHandleChange: fieldNames => validators => value => {
    const formId = props.form.getFormId();
    const newValue = props.normalize && props.normalize(value) || value;
    const errorMessages = createValidate(validators)(newValue);
    if (errorMessages.length === 0) {
      dispatch(setValue(formId, fieldNames, newValue));
    } else {
      dispatch(mergeFields(formId, [{
        fieldName: fieldNames,
        object: {
          errorMessages,
          value: newValue,
          touched: true,
        },
      }]));
    }
    props.changeHandlers && props.changeHandlers.forEach(changeHandler => changeHandler(newValue));
  },
});

export default compose(
  withForm,
  applyReducers(reducers),
  connect(mapStateToProps, mapDispatchToProps),
  _createField
);
