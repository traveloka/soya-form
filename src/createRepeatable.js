import { compose } from "redux";
import { connect } from "react-redux";
import { applyReducers } from "soya-next/redux";
import reducers from "./_reducers";
import {
  addListItem,
  removeListItem,
  reorderListItem,
  reorderListItemDec,
  reorderListItemInc
} from "./_actions";
import createSelector from "./_selectors";
import { getFieldNames } from "./_utils";
import withForm from "./withForm";
import _createRepeatable from "./_createRepeatable";

const mapStateToProps = (state, props) => {
  const fieldNames = getFieldNames(props.name);
  const selector = createSelector(state, props.form.getFormId());
  const length = selector.getLength(fieldNames);
  const minLength =
    typeof props.minLength === "number" && props.minLength >= 0
      ? props.minLength
      : 1;

  return {
    length: length >= minLength ? length : minLength,
    name: fieldNames
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  addListItem: fieldNames => () => {
    dispatch(
      addListItem(
        props.form.getFormId(),
        fieldNames,
        props.minLength,
        props.maxLength
      )
    );
  },
  removeListItem: fieldNames => index => {
    dispatch(removeListItem(props.form.getFormId(), fieldNames, index));
  },
  reorderListItem: fieldNames => (index, targetIndex) => {
    dispatch(
      reorderListItem(props.form.getFormId(), fieldNames, index, targetIndex)
    );
  },
  reorderListItemDec: fieldNames => (index, amount) => {
    dispatch(
      reorderListItemDec(props.form.getFormId(), fieldNames, index, amount)
    );
  },
  reorderListItemInc: fieldNames => (index, amount) => {
    dispatch(
      reorderListItemInc(props.form.getFormId(), fieldNames, index, amount)
    );
  }
});

export const createRepeatableAction = Component =>
  compose(
    withForm,
    applyReducers(reducers),
    connect(
      mapStateToProps,
      mapDispatchToProps
    )
  )(Component);

export default compose(
  withForm,
  applyReducers(reducers),
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  _createRepeatable
);
