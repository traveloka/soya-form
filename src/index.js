import createForm from "./createForm";
import createField from "./createField";
import createRepeatable from "./createRepeatable";
import createSelector from "./_selectors";
import withForm from "./withForm";
import withFormValues from "./withFormValues";
import { transform as getSoyaFormFormat } from "./getSoyaFormFormat";

export {
  createForm,
  createField,
  createRepeatable,
  createSelector,
  withForm,
  withFormValues,
  getSoyaFormFormat
};
