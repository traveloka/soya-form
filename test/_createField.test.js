import React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import { __createForm } from '../src/_createForm';
import _createField from '../src/_createField';

describe('field container', () => {
  class Field extends React.Component {
    render() {
      return <div />;
    }
  }
  const reduceMockProps = value => (props, name) => ({
    ...props,
    [name]: value,
  });
  const formId = 'simpleForm';
  const FieldContainer = _createField(Field);
  let fields, fieldNames, renderer;
  let props = {
    errorMessages: [],
    name: ['foo'],
    ...[
      'isFormEnabled',
      'isFormSubmitting',
      'isDisabled',
      'isFieldEnabled',
      'isValidating',
      'touched',
    ].reduce(reduceMockProps(false), {}),
    ...[
      'setDefaultValue',
      'addErrorMessages',
      'mergeFields',
      'toggleLock',
    ].reduce(reduceMockProps(() => jest.fn()), {}),
    createHandleChange: () => () => jest.fn(),
  };

  beforeEach(() => {
    fields = {};
    fieldNames = [];
    renderer = createRenderer();
    props = {
      ...props,
      form: __createForm(fields, fieldNames)(formId, action => action),
      asyncValidators: [jest.fn(), jest.fn(), jest.fn()],
      changeValidators: [jest.fn(), jest.fn(), jest.fn()],
      submitValidators: [jest.fn(), jest.fn(), jest.fn()],
    };
    renderer.render(<FieldContainer {...props} />);
  });

  it('should create field component', () => {
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it('should handle register async validators', () => {
    const fieldContainer = renderer.getMountedInstance();
    fieldContainer.registerAsyncValidators([jest.fn(), jest.fn(), jest.fn()]);
    expect(fieldContainer.__asyncValidators).toMatchSnapshot();
  });

  it('should handle register change validators', () => {
    const fieldContainer = renderer.getMountedInstance();
    fieldContainer.registerChangeValidators([jest.fn(), jest.fn(), jest.fn()]);
    expect(fieldContainer.__changeValidators).toMatchSnapshot();
  });

  it('should handle register submit validators', () => {
    const fieldContainer = renderer.getMountedInstance();
    fieldContainer.registerSubmitValidators([jest.fn(), jest.fn(), jest.fn()]);
    expect(fieldContainer.__submitValidators).toMatchSnapshot();
  });

  it('should register change, async, and submit validators on mount', async () => {
    await fields[props.name].validateAll();
    expect([
      ...props.asyncValidators,
      ...props.changeValidators,
      ...props.submitValidators,
    ].reduce((sum, { mock }) => sum + mock.calls.length, 0)).toBe(9);
  });

  it('should unregister change, async, and submit validators on unmount', () => {
    renderer.unmount();
    expect({
      __fields: fields,
      __fieldNames: fieldNames,
    }).toMatchSnapshot();
  });

  it('should return value correctly', async () => {
    renderer.render(<FieldContainer {...props} value={'my value'} />);

    expect(await fields[props.name].validateAll()).toEqual({
      isValid: true,
      name: ['foo'],
      value: 'my value'
    });
  });

  it('should throw error parseValue is not a function', async () => {
    renderer.render(<FieldContainer {...props} value='test' parseValue={'test'} />);
    await expect(fields[props.name].validateAll()).rejects.toEqual(expect.any(Error))
  });

  it('should call parseValue correctly', async () => {
    const parseValue = jest.fn().mockImplementation(value => value.id);
    renderer.render(<FieldContainer {...props} value={{ id: '1234', text: 'Testing' }} parseValue={parseValue} />);

    expect(await fields[props.name].validateAll()).toEqual({
      isValid: true,
      name: ['foo'],
      value: '1234'
    });
    expect(parseValue).toBeCalled();
  });
});
