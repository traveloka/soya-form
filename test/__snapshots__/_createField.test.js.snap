// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field container should create field component 1`] = `
<Field
  errorMessages={Array []}
  form={
    Object {
      "addErrors": [Function],
      "clearErrors": [Function],
      "clearForm": [Function],
      "disable": [Function],
      "disableField": [Function],
      "disableFields": [Function],
      "enable": [Function],
      "enableField": [Function],
      "enableFields": [Function],
      "getFormId": [Function],
      "lockSubmission": [Function],
      "regField": [Function],
      "setDefaultValue": [Function],
      "setDefaultValues": [Function],
      "setErrors": [Function],
      "setValue": [Function],
      "setValues": [Function],
      "submit": [Function],
      "unlockSubmission": [Function],
      "unregField": [Function],
      "validateAll": [Function],
    }
  }
  handleAsyncValidation={[Function]}
  handleChange={[Function]}
  isDisabled={false}
  isFieldEnabled={false}
  isFormEnabled={false}
  isFormSubmitting={false}
  isValidating={false}
  name={
    Array [
      "foo",
    ]
  }
  parseValue={[Function]}
  registerAsyncValidators={[Function]}
  registerChangeValidators={[Function]}
  registerSubmitValidators={[Function]}
  setDefaultValue={[Function]}
  touched={false}
/>
`;

exports[`field container should handle register async validators 1`] = `
Array [
  [Function],
  [Function],
  [Function],
]
`;

exports[`field container should handle register change validators 1`] = `
Array [
  [Function],
  [Function],
  [Function],
]
`;

exports[`field container should handle register submit validators 1`] = `
Array [
  [Function],
  [Function],
  [Function],
]
`;

exports[`field container should unregister change, async, and submit validators on unmount 1`] = `
Object {
  "__fieldNames": Array [],
  "__fields": Object {},
}
`;
